Clem.tmdialects upstreamable

ljfitz opened the pull request:
This follows up on #623 for out-of-tree builds of torch-mlir, which added building `torch-mir-dialects` as a subdirectory.

We found that we needed some more `cmake` adaptions to make this work. We're initially opening this PR as a draft to gather opinions as to whether this approach suits at least the short-term `torch-mlir-dialects` goals.

Our goal is to support both in-tree and out-of-tree builds of `torch-mlir` with minimum hassle, for instance by using the same variable names in both setups. Specifically in `externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt` :
- We use `MLIR_FOUND` to detect that it is being build as a subdirectory and the llvm+mlir cmake infrastructure is already set up (via find_package in the parent build) as opposed to an in-tree build.
- For in-tree,  the setting of variables and loading of llvm+mlir cmake infrastructure is now conditionally performed.
- For in-tree, the names of cmake variables being defined for are adjusted to match those `llvm-project` makes available through `find_package(MLIR REQUIRED CONFIG)`, under the assumption that those are the more "standardized" names.

Further optional items we could contribute if you're interested:
- build instructions for an out-of-tree setup (eg in README.md)
- a CI flow to test an out-of-tree setup, so that all code paths of the cmake scripts are tested (even though we expect most llvm developers to use in-tree builds)

ljfitz committed:
Fix out-of-tree build of torch-mlir-dialects

Follows up on #623 for out-of-tree builds of torch-mlir, which
added building `torch-mir-dialects` as a subdirectory.

Our goal is to support both in-tree and out-of-tree builds of
`torch-mlir` with minimum hassle, for instance by using the same
variable names in both setups.

Specific changes to `externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt`:
- We use `MLIR_FOUND` to detect that it is being build as a subdirectory
and the llvm+mlir cmake infrastructure is already set up (via
find_package in the parent build) as opposed to an in-tree build.
- For in-tree, the setting of variables and loading of llvm+mlir cmake
infrastructure is now conditionally performed.
- For in-tree, the names of cmake variables being defined for are
adjusted to match those `llvm-project` makes available through
`find_package(MLIR REQUIRED CONFIG)`, under the assumption that those
are the more "standardized" names.

Co-authored-by: Clément Fournier <clement.fournier@amd.com>

silvasean committed:
Fix for 0-size dim inferred incorrectly.

The issue was in the canonicalizer for torch.aten.ge.int -- in cases
where the operands were swapped, it would miscompile. This issue is
fixed and folding support generalized to `torch.aten.size.int < 0` as
well.

Fixes #716

anupgangwar committed:
tosa] Support for Aten[Gelu|GeluBackward] ops (#720)

Signed-off-by: Anup Gangwar <anup.gangwar@arm.com>

Co-authored-by: Anup Gangwar <anup.gangwar@arm.com>

ljfitz commented on externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt:
```diff
@@ -1,30 +1,36 @@
-  "This project is intended to be built as part of LLVM via "
-  "-DLLVM_EXTERNAL_PROJECTS=torch-mlir-dialects "
-  "-DLLVM_EXTERNAL_TORCH_MLIR_DIALECTS_SOURCE_DIR=${CMAKE_CURRENT_SOURCE_DIR}")
+  set(LLVM_OUT_OF_TREE_BUILD TRUE)
```
Put this back as an error condition - otherwise we'd need to introduce most of the out-of-tree setup from torch's cmake.

ljfitz commented on CMakeLists.txt:
```diff
@@ -36,13 +36,13 @@ macro(torch_mlir_add_llvm_external_project name identifier location)
 
 if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
   # Out-of-tree build
+  set(LLVM_OUT_OF_TREE_BUILD TRUE)
```
Maybe `BUILD_TORCH_MLIR_DIALECTS_AS_SUBDIRECTORY` and move to just before the `add_subdirectory`.

ljfitz commented on CMakeLists.txt:
```diff
@@ -36,13 +36,13 @@ macro(torch_mlir_add_llvm_external_project name identifier location)
-  torch-mlir-dialects
-  TORCH_MLIR_DIALECTS
-  ${CMAKE_CURRENT_SOURCE_DIR}/externals/llvm-external-projects/torch-mlir-dialects)
+set(TORCH_MLIR_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
```
This can go back to where it was.

ljfitz commented on CMakeLists.txt:
```diff
@@ -82,7 +82,14 @@ if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
   set(BACKEND_PACKAGE_STRING "LLVM ${LLVM_PACKAGE_VERSION}")
   add_subdirectory(externals/llvm-external-projects/torch-mlir-dialects)
 else()
+  set(LLVM_OUT_OF_TREE_BUILD FALSE)
```
Lets leave cmake "undefined is false" for `BUILD_TORCH_MLIR_DIALECTS_AS_SUBDIRECTORY `.

ljfitz commented on externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt:
```diff
@@ -1,30 +1,36 @@
 set(TORCH_MLIR_DIALECTS_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
 set(TORCH_MLIR_DIALECTS_BINARY_DIR "${CMAKE_CURRENT_BINARY_DIR}")
 message(STATUS "Building torch-mlir-dialects project at ${TORCH_MLIR_DIALECTS_SOURCE_DIR} (into ${TORCH_MLIR_DIALECTS_BINARY_DIR})")
+option(MLIR_ENABLE_BINDINGS_PYTHON "Enables MLIR Python Bindings" OFF)
```
Put back where it was.

ljfitz commented on externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt:
```diff
@@ -1,30 +1,36 @@
+  message(STATUS "torch-mlir-dialect is being built in-tree")
+  # An in-tree build, LLVM hasn't been installed yet. Rebuild LLVM paths.
+  set(MLIR_MAIN_SRC_DIR ${LLVM_MAIN_SRC_DIR}/../mlir)
+  set(MLIR_INCLUDE_DIR ${LLVM_BINARY_DIR}/tools/mlir/include)
```
# Create the same variables as `find_package MLIR ...` defines.

ljfitz commented on externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt:
```diff
@@ -1,30 +1,36 @@
   set(_dirs
-    $<BUILD_INTERFACE:${MLIR_INCLUDE_DIR}>
-    $<BUILD_INTERFACE:${MLIR_GENERATED_INCLUDE_DIR}>
+    $<BUILD_INTERFACE:${LLVM_BUILD_MAIN_SRC_DIR}/include>
```
Check which file is needed from `llvm/include`?

ljfitz commented on CMakeLists.txt:
```diff
@@ -108,7 +114,7 @@ include_directories(${CMAKE_CURRENT_BINARY_DIR}/include)
 function(torch_mlir_target_includes target)
   set(_dirs
-    $<BUILD_INTERFACE:${MLIR_INCLUDE_DIR}>
+    $<BUILD_INTERFACE:${MLIR_INCLUDE_DIRS}>
```
I was surprised to see this change. Does it suggest we should also remove the `MLIR_GENERATED_INCLUDE_DIR` from this file too?

oowekyala replied:
Yes you're right!

ljfitz commented on externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt:
```diff
@@ -1,4 +1,7 @@
 if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+# This disables the setup whereby torch-mlir-dialects is
+# built as a completely standalone project, ie neither as
+# part of LLVM or as a subdirectory of torch-mlir.
```
```suggestion
# The cmake configuration to build torch-mlir-dialects on
# its own out-of-tree has not been implemented. It can
# be built as part of LLVM or as a subdirectory of torch-mlir.
```

ljfitz commented on CMakeLists.txt:
```diff
@@ -80,9 +76,19 @@ if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+  set(BUILD_TORCH_MLIR_DIALECTS_AS_SUBDIRECTORY TRUE)
   add_subdirectory(externals/llvm-external-projects/torch-mlir-dialects)
 else()
+
```
```suggestion
```

oowekyala commented on externals/llvm-external-projects/torch-mlir-dialects/CMakeLists.txt:
```diff
@@ -11,20 +14,34 @@ set(TORCH_MLIR_DIALECTS_SOURCE_DIR "${CMAKE_CURRENT_SOURCE_DIR}")
   set(_dirs
-    $<BUILD_INTERFACE:${MLIR_INCLUDE_DIR}>
-    $<BUILD_INTERFACE:${MLIR_GENERATED_INCLUDE_DIR}>
+    $<BUILD_INTERFACE:${LLVM_INCLUDE_DIRS}>
```
I'm not entirely sure we need this now

oowekyala replied:
I've rempved it

ramiro050 committed:
Add support for multi-dim input to `index_put_impl` (#722)

This commit adds support for multi-dimensional tensors as input to the
`_index_put_impl_` op. The support was to some degree already there,
since `ScatterOp` already supports multi-dimensional tensors. This
commit also adds a bit more error checking to `index_put` and
refactors the code for creating `ScatterOp`s to mimic the way one
would make a `Linalg::GenericOp`.

oowekyala commented:
Was merged upstream llvm/torch-mlir#726
